module/zfs/vdev.c

```diff
@@ -3466,6 +3453,12 @@ vdev_set_state(vdev_t *vd, boolean_t isopen, vdev_state_t state, vdev_aux_t aux)
 		vd->vdev_removed = B_FALSE;
 	}
 
+	/*
+	 * Notify ZED of any state change to a leaf vdev
+	 */
+	if (vd->vdev_ops->vdev_op_leaf)
+		zfs_post_state_change(spa, vd, save_state);
```

Comment:
I think the naming here is going to be downright confusing since we already have a `vdev->vdev_prevstate` member which means something slightly different.

How about we rename the existing `vdev->vdev_prevstate` to `vdev->vdev_closestate` and then use the now available `vdev->vdev_prevstate` for `save_state`.  Then the FM_EREPORT_PAYLOAD_ZFS_VDEV_PREVSTATE key wouldn't be misleading when reading the code.  And you wouldn't need to change the `zfs_post_state_change()`  prototype.